src/Authoring/WinRT.SourceGenerator/DiagnosticHelpers.cs

```diff
@@ -0,0 +1,365 @@
+        {
+            Flag();
+            _context.ReportDiagnostic(Diagnostic.Create(d, loc, args));
+        }
+
+        private void ReportDiagnostic(Diagnostic d)
+        {
+            Flag();
+            _context.ReportDiagnostic(d);
+        }
+
+        ///<summary>Array types can only be one dimensional and not System.Array, 
+        ///and there are some types not usable in the Windows Runtime, like KeyValuePair</summary> 
+        ///<param name="typeSymbol">The type to check</param><param name="loc">where the type is</param>
+        ///<param name="memberId">The method or property with this type in its signature</param>
+        /// <param name="typeId">the type this member (method/prop) lives in</param>
+        private void ReportIfInvalidType(ITypeSymbol typeSymbol, Location loc, SyntaxToken memberId, SyntaxToken typeId) 
+        { 
+
+            if (typeSymbol.TypeKind == TypeKind.Array) 
+            { 
+                IsInvalidArrayType((IArrayTypeSymbol)typeSymbol, loc, memberId, typeId);
+                return;
+            }
+
+            foreach (var s in InvalidTypes)
+            { 
+                var invalidTypeSym = _context.Compilation.GetTypeByMetadataName(s);
+                if (SymbolEqualityComparer.Default.Equals(typeSymbol.OriginalDefinition, invalidTypeSym))
+                {
+                    Report(WinRTRules.UnsupportedTypeRule, loc, memberId, s, SuggestType(s));
+                    return;
+                }
+            }
+
+            // helper function, we can't get a type symbol for some of the invalid types 
+            bool SameType(ITypeSymbol sym, string s)
+            {
+                string baseStr = sym.ContainingNamespace.IsGlobalNamespace ? "" : sym.ContainingSymbol + ".";
+                var x = baseStr + sym.MetadataName;
+                return x == s;
+            }
+
+            // GetTypeByMetadataName fails on "System.Linq.Enumerable`1" & "System.Collections.ObjectModel.ReadOnlyDictionary`2"
+            // Would be fixed by issue #678 on the dotnet/roslyn-sdk repo
+            foreach (var invalidType in WIPInvalidTypes) 
+            {
+                if (SameType(typeSymbol, invalidType))
+                { 
+                    Report(WinRTRules.UnsupportedTypeRule, loc, memberId, invalidType, SuggestType(invalidType));
+                    return;
+                }
+            }
+        }
+ 
+        /// <summary>
+        /// See if this class/interfaces inherits the given type
+        /// </summary>
+        /// <param name="typeSymbol">type that might inherit</param>
+        /// <param name="typeToCheck">Inherited interface or class</param>
+        private bool ImplementsInterface(INamedTypeSymbol typeSymbol, string typeToCheck)
+        {
+            if (typeSymbol == null)
+            {
+                return false;
+            }
+
+            if (typeSymbol.BaseType != null)

```

Comment:
Technically the base type can have another base type.  Does that matter in the scenarios this is used in?

Reply:
It seems like there is some ambiguity about the `BaseType` field of type symbols. For type symbols of classes, any interface implemented is actually a basetype (e.g. IAsyncAction is not in the symbol's `AllInterfaces` field, but in its `BaseType` field).

For an interface type symbol, this is not true, and the `IAsyncAction` symbol **is** in the `AllInterfaces` field. So the loop is for interfaces, when we want to check what a class implements, we need it's `BaseType`, but an interface has a null `BaseType`. 

And class type symbols need a `BaseType` check, to find `System.Exception` for example. 

It is also curious that calling `GetTypeByMetadataName("Windows.Foundation.IAsyncAction")` returns `null` instead of the type symbol we might expect. This is why we have to do the `==` check.